server/src/main/java/org/apache/druid/segment/realtime/appenderator/BatchAppenderator.java

```diff
@@ -785,6 +785,9 @@ private DataSegment mergeAndPush(
         for (FireHydrant fireHydrant : sink) {
           Pair<ReferenceCountingSegment, Closeable> segmentAndCloseable = fireHydrant.getAndIncrementSegment();
           final QueryableIndex queryableIndex = segmentAndCloseable.lhs.asQueryableIndex();
+          if (queryableIndex != null) {
+            metrics.incrementMergeRows(queryableIndex.getNumRows());
```

Comment:
This isn't the right place to increment the number of rows merged; it's happening _before_ merge so it's really the number of rows that are about-to-be-merged. If you want the number of rows actually merged, the hook point should be after `mergeQueryableIndex` is done.